### PR TITLE
[flang][OpenMP] Add support for complex reductions

### PR DESCRIPTION
**Pull request by Leporacanthicus**

### CODE REVIEW

File: flang/lib/Lower/OpenMP/ReductionProcessor.cpp [hunk 1]
```diff
@@ -160,6 +176,17 @@ ReductionProcessor::getReductionInitValue(mlir::Location loc, mlir::Type type,
...
+    if (auto cplxTy = mlir::dyn_cast<fir::ComplexType>(type)) {
+      mlir::Type realTy =
+          Fortran::lower::convertReal(builder.getContext(), cplxTy.getFKind());
+      //      mlir::FloatType realTy =
+      //      mlir::dyn_cast<mlir::FloatType>(cplxTy.getElementType());
+      //      const llvm::fltSemantics &sem = (realTy).getFloatSemantics();
+      mlir::Value init = builder.createRealConstant(
+          loc, realTy, getOperationIdentity(redId, loc));
+      return fir::factory::Complex{builder, loc}.createComplex(type, init,
+                                                               init);
```
Comment (tblah) [inline flang/lib/Lower/OpenMP/ReductionProcessor.cpp:188]: The OpenACC code always initializes the imaginary part with `0.0`. What does the standard say about this?

**Review by tblah — COMMENTED** [review]
Do we support passing complex by val or is this only for by-ref reductions? IIRC we require by-ref reductions to have a trivial type so I guess we don't support that, but it isn't obvious from this patch

**Review by tblah — APPROVED** [review]
Thanks for this, LGTM (but I see a merge conflict)

**Comment by tblah** [comment]
> Do we support passing complex by val or is this only for by-ref reductions? IIRC we require by-ref reductions to have a trivial type so I guess we don't support that, but it isn't obvious from this patch

I can see from the tests that we do support by-value complex reductions so ignore this comment

**Comment by ye-luo** [comment]
Verified. It resolves https://github.com/llvm/llvm-project/issues/87839 and the results are correct.

**Pull request merged**